packages/pigeon/lib/cpp_generator.dart

```diff
@@ -171,6 +171,220 @@ template<class T> class ErrorOr {
+        root.enums, (NamedType x) => _baseCppTypeForBuiltinDartType(x.type));
+    final String instanceVariableName = _makeInstanceVariableName(field);
+    final String qualifiedGetterName =
+        '${klass.name}::${_makeGetterName(field)}';
+    final String qualifiedSetterName =
+        '${klass.name}::${_makeSetterName(field)}';
+    final String returnExpression = hostDatatype.isNullable
+        ? '$instanceVariableName ? &(*$instanceVariableName) : nullptr'
+        : instanceVariableName;
+
+    // Generates the string for a setter treating the type as [type], to allow
+    // generating multiple setter variants.
+    String makeSetter(HostDatatype type) {
+      const String setterArgumentName = 'value_arg';
+      final String valueExpression = type.isNullable
+          ? '$setterArgumentName ? ${_valueType(type)}(*$setterArgumentName) : std::nullopt'
+          : setterArgumentName;
+      return 'void $qualifiedSetterName(${_unownedArgumentType(type)} $setterArgumentName) '
+          '{ $instanceVariableName = $valueExpression; }';
+    }
+
+    indent.writeln(
+        '${_getterReturnType(hostDatatype)} $qualifiedGetterName() const '
+        '{ return $returnExpression; }');
+    indent.writeln(makeSetter(hostDatatype));
+    if (hostDatatype.isNullable) {
+      // Write the non-nullable variant; see _writeDataClassDeclaration.
+      final HostDatatype nonNullType = _nonNullableType(hostDatatype);
+      indent.writeln(makeSetter(nonNullType));
+    }
+
+    indent.addln('');
+  }
+
+  // Serialization.
+  indent.write('flutter::EncodableMap ${klass.name}::ToEncodableMap() const ');
+  indent.scoped('{', '}', () {
+    indent.scoped('return flutter::EncodableMap{', '};', () {
+      for (final NamedType field in klass.fields) {
+        final HostDatatype hostDatatype = getHostDatatype(
+            field,
+            root.classes,
+            root.enums,
+            (NamedType x) => _baseCppTypeForBuiltinDartType(x.type));
+
+        final String instanceVariable = _makeInstanceVariableName(field);
+
+        final String encodableKey = 'flutter::EncodableValue("${field.name}")';
+        String encodableValue = '';
+        if (!hostDatatype.isBuiltin &&
+            rootClassNameSet.contains(field.type.baseName)) {
+          final String operator = field.type.isNullable ? '->' : '.';
+          encodableValue = '$instanceVariable${operator}ToEncodableMap()';
+        } else if (!hostDatatype.isBuiltin &&
+            rootEnumNameSet.contains(field.type.baseName)) {
+          final String nonNullValue =
+              field.type.isNullable ? '(*$instanceVariable)' : instanceVariable;
+          encodableValue = 'flutter::EncodableValue((int)$nonNullValue)';
+        } else {
+          final String operator = field.type.isNullable ? '*' : '';
+          encodableValue =
+              'flutter::EncodableValue($operator$instanceVariable)';
+        }
+
+        if (field.type.isNullable) {
+          encodableValue =
```

Comment:
Shouldn't this logic be changed so we generate an `if(....has_value())` then add to the map, else, doesn't? It would need to be single `.insert()`s instead of an array initializer, but I think it would make sense, no? Or adding the `EncodableValue()` is pretty much the same as doing that?

Reply:
In theory we could do that, but it would be inconsistent with what the existing generators do which makes me a bit nervous as the runtime behavior is subtly different. I filed https://github.com/flutter/flutter/issues/105233 for follow-up.

Reply:
Let's deal with it on the new issue then. I'll mark this as resolved.

Reply:
SG. Any other feedback on the PR?

Reply:
Still going through it. I'll finish the review today.